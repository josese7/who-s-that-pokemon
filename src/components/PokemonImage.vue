<template>
    <div class="pokemon-container">
        <img   v-if="!showPokemon" class= "pokemon-hidden" :src="imgSrc" alt="">
        <img  v-else class="fade-in" :src="imgSrc" alt="">
        

    </div>
</template>

<script>
export default {
    props:{
        pokemonId:{
            type: Number,
            require:true

        },
        showPokemon:{
            type: Boolean,
            required: true
        }
    },
    computed:{
        imgSrc() {
            return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/${this.pokemonId}.svg`
        }
    }
}
</script>

<style scoped>
.pokemon-container {
    height: 20px;
}
img {
    height: 200px;
    position: absolute;
    left: 0;
    right: 0;
    margin: 0 auto; 
    user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    -webkit-user-drag: none;
    -webkit-user-select: none;
}
.pokemon-hidden {
    filter: brightness(0);
}
</style>