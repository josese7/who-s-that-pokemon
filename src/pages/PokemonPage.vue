<template>
      <h1 v-if="!pokemon">Espere un momento...</h1>
      <div v-else>
        <h1>¿Quien es este pokemon?</h1>
        <PokemonImage  :pokemon-id="pokemon.id" :showPokemon="showPokemon"/>
        <PokemonOptions 
        :pokemons="pokemonsArray"
        @selection="checkAnswer"

        />
        <template v-if="showAnswer">
          <h2>{{message}}</h2>
          <button
          @click="newGame"
          >Jugar de nuevo</button>
        </template>
        
      </div>
     

 
</template>

<script>

import PokemonImage from '../components/PokemonImage.vue'
import PokemonOptions from '../components/PokemonOptions.vue'

import getPokemonsOptions from '@/helpers/getPokemonsOptions.js'



export default {
  components: {
    PokemonImage,
    PokemonOptions
  },
  data(){
    return{
      pokemonsArray: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message:''
   
    }
  },
  methods:{
    async mixPokemonArray() {
      this.pokemonsArray = await getPokemonsOptions()
      //console.log(this.pokemonsArray)
      const  randPokemon = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonsArray[randPokemon]
      //console.log(this.pokemon)
    },
    checkAnswer(pokemonId){

      this.showPokemon = true
      this.showAnswer = true
      this.message= pokemonId === this.pokemon.id? `Correcto, es ${this.pokemon.name}.¡Estás cada vez más cerca de ser un MAESTRO POKEMON!`
      : `Oops, lo siento! Era ${this.pokemon.name}`;
      return
    },
    async newGame(){
  
      await this.mixPokemonArray()
      this.showPokemon= false
      this.showAnswer=false
      this.message=''
    }
  },
  mounted(){
    this.mixPokemonArray()
  }

}
</script>

<style>

</style>