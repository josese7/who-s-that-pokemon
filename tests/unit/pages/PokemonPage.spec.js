import { mount, shallowMount } from "@vue/test-utils";
import PokemonPage from '@/pages/PokemonPage'
import { pokemonsMock } from "../mocks/pokemons.mock";
import PokemonImage from '@/components/PokemonImage'
import PokemonOptions from '@/components/PokemonOptions'




describe('PokemonPage Page', ()=>{

    let wrapper
    let mixPokemonArraySpy

    beforeEach(()=>{
        mixPokemonArraySpy= jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
        wrapper = shallowMount(PokemonPage)

        
    })
    afterEach(()=>{
        jest.clearAllMocks()
    })
    test('snapshot inicial', () => { 
        expect(wrapper.html()).toMatchSnapshot()
     })

     test('debe de llamarse a mixPokemonArr al montar el componente', ()=>{
        expect(mixPokemonArraySpy).toHaveBeenCalled()
     })
     
     test('snapshot con opciones de pokemin', ()=>{

        wrapper = shallowMount(PokemonPage, {
            data(){
                return{
                    pokemonsArray: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message:''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()

     })

     test('deben de existir PokemonImage y PokemonsOptions', () => { 

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ""
                }
            }
        })

        //[pokemonImage, pokemonsOptions ]= wrapper.findAllComponents()
        
        const image = wrapper.findComponent(PokemonImage)
        const options = wrapper.findComponent(PokemonOptions)
        // image = wrapper.find('pokemon-image-stub')
        // image = wrapper.find('pokemon-options-stub')
        const { pokemonId } = image.props()
        const { pokemons } = options.props()
       
        expect(image.exists()).toBeTruthy()
        expect(options.exists()).toBeTruthy() 

        expect(pokemonId).toBe(1)
        expect(wrapper.find('pokemon-options-stub').exists()).toBeTruthy()
        /* 
                no funciona
        expect(pokemons).toEqual(
            [
                {
                    name: expect.any(String),
                    id: expect.any(Number)
                },
                {
                    name: expect.any(String),
                    id: expect.any(Number)
                },
                {
                    name: expect.any(String),
                    id: expect.any(Number)
                },
                {
                    name: expect.any(String),
                    id: expect.any(Number)
                },
            ]
        ) */
      })

      test('Probando checkAnswer', async ()=>{

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemonsMock,
                    pokemon: pokemonsMock[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ""
                }
            }
        })
        //Es una peticion asincrona, se debe esperar a que Vue re-renderize los cambios de las propiedades reactivas
        await wrapper.vm.checkAnswer(1)

        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`Correcto, es ${pokemonsMock[0].name}.¡Estás cada vez más cerca de ser un MAESTRO POKEMON!`)

        await wrapper.vm.checkAnswer(67)

        expect(wrapper.find('h2').text()).toBe(`Oops, lo siento! Era ${pokemonsMock[0].name}`) 




      })
})