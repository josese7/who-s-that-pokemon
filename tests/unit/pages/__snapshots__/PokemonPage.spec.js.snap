// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Page snapshot con opciones de pokemin 1`] = `
<div>
  <h1>Â¿Quien es este pokemon?</h1>
  <pokemon-image-stub pokemonid="1" showpokemon="false"></pokemon-image-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonPage Page snapshot inicial 1`] = `<h1>Espere un momento...</h1>`;
